File: Homework2/Codes/22202709_Ozongun_HalilArda.go
package main

import "fmt"


type Obj struct {
	x int
}

func main() {
	fmt.Println("1 - What is the form and type of the control expression?")

	// Control expression as a variable
	value := 5
	switch value {
	case 1:
		fmt.Println("Control expression is an integer")
	case 5:
		fmt.Println("Control expression is an integer")
	}

	// Control expression as an expression
	val1 := 3
	val2 := 4
	switch val1 + val2 {
	case 7:
		fmt.Println("Control expression is an expression resulting in 7")
	}

	// Control expression as a literal
	switch 10 {
		case 10:
		fmt.Println("Control expression is a literal")
		case 20:
		fmt.Println("Control expression is a literal")

	}

	// Control expression as a object
	obj := Obj{4}
	switch obj {
		case Obj{3}:
			fmt.Println("Object with x = 3")
		case Obj{4}:
			fmt.Println("Object with x = 4")
	}

	// Type switch
	var i interface{}
	items := []interface{}{1, "apple", [2]int{3, 4}, []int{1, 2, 3}, map[string]int{"key": 1}, true, nil, Obj{3}}

	for _, i = range items {
		// print value and type of i
		fmt.Printf("Value: %v, Type: %T\t\t", i, i)
		switch v := i.(type) {
		case int:
			fmt.Println("Control expression is an integer")
		case string:
			fmt.Println("Control expression is a string")
		case [2]int:
			fmt.Println("Control expression is an array length 2")
		case []int:
			fmt.Println("Control expression is a slice")
		case map[string]int:
			fmt.Println("Control expression is a map")
		case bool:
			fmt.Println("Control expression is a boolean")
		case nil:
			fmt.Println("Control expression is nil")
		case Obj:
			if v.x == 3 {
				fmt.Println("Control expression is an Obj with x = 3")
			} else {
				fmt.Println("Control expression is an Obj")
			}
		default:
			fmt.Println("No match found")
		}
	}




	fmt.Println("\n\n2 - How are the selectable segments specified?")
	value = 5
	switch value {
	case 5:
		fmt.Println("Case 5: Single statement")
	case 10:
		fmt.Println("Case 10: Multiple statements")
		fmt.Println("You can have more than one line in a case block")
	}



	fmt.Println("\n\n3 - Is execution flow restricted to just a single selectable segment?")
	value = 6
	switch value {
	case 2, 4, 6:
		fmt.Println("Matched even number")
	case 3, 9: 
		fmt.Println("Matched divisible by 3")
	default:
		fmt.Println("No match")
	}
	value = 6
	switch value {
	case 2, 4, 6:
		fmt.Println("Matched even number")
		fallthrough
	case 3, 9: 
		fmt.Println("Matched divisible by 3")
	default:
		fmt.Println("No match")
	}
	/*switch value {
	case 2, 4, 6:
		fmt.Println("Matched even number")
		case 3, 6, 9:
		fmt.Println("Matched divisible by 3")
	}*/




	fmt.Println("\n\n4 - How are case values specified?")

	value = 3
	// literal values
	switch value {
	case 1:
		fmt.Println("Matched 1")
	case 2:
		fmt.Println("Matched 2")
	case 3:
		fmt.Println("Matched 3")
	}

	// expressions
	switch value {
	case 2 + 1:
		fmt.Println("Matched 3")
	case 10 - 3:
		fmt.Println("Matched 7")
	}

	// expression with variables
	var m, n int = 2, 3
	switch value {
	case m * n:
		fmt.Println("Matched 6")
	case m + n:
		fmt.Println("Matched 5")
	}

	// multiple case values

	switch value {
	case 1, 2, 3:
		fmt.Println("Matched 1, 2, or 3")
	case 4, 5, 6:
		fmt.Println("Matched 4, 5, or 6")
	}

	// variables
	var a, b int = 1, 3
	switch value {
	case a:
		fmt.Println("Matched a")
	case b:
		fmt.Println("Matched b")
	}

	// different types not allowed: 
	// "Compile time error: cannot convert "2" (untyped string constant) to type int"
	/*switch value {
	case 1:
		fmt.Println("Matched 1")
	case "2":
		fmt.Println("Matched string 2")
	}*/

	// type switch
	var i2 interface{} = 4
	switch i2.(type) {
	case int:
		fmt.Println("Matched int")
	case string:
		fmt.Println("Matched string")
	case bool:
		fmt.Println("Matched bool")
	case nil:
		fmt.Println("Matched nil")
	}
	
	


	fmt.Println("\n\n5 - What is done about unrepresented expression values?")
	value = 100
	switch value {
	case 10:
		fmt.Println("Matched 10")
	default:
		fmt.Println("No match for the given value")
	}
}

File: Homework2/Codes/22202709_Ozongun_HalilArda.rs
fn main() {
    println!("1 - What is the form and type of the control expression?");

    // variable
    let value = 5;
    match value {
        1 => println!("Control expression is an integer"),
        5 => println!("Control expression is an integer"),
        _ => (),
    }

    // expression
    let val1 = 3;
    let val2 = 4;
    match val1 + val2 {
        7 => println!("Control expression is an expression"),
        _ => (),
    }

    // literal
    match 10 {
        10 => println!("Control expression is a literal"),
        20 => println!("Control expression is a literal"),
        _ => (),
    }

    // object
    #[derive(PartialEq, Debug)]
    struct Obj {
        x: i32,
    }
    let obj = Obj { x: 4 };
    match obj {
        Obj { x: 3 } => println!("Object with x = 3"),
        Obj { x: 4 } => println!("Object with x = 4"),
        _ => (),
    }

    // Option type
    let some_value: Option<i32> = Some(5);
    match some_value {
        Some(n) => println!("Got a value: {}", n),
        None => println!("Got None"),
    }

    // Match on different types using enums
    println!("\nExample with different types using enums:");
    
    #[derive(Debug)]
    enum Item {
        Integer(i32),
        Text(String),
        Array([i32; 2]),
        Vector(Vec<i32>),
        Map(std::collections::HashMap<String, i32>),
        Boolean(bool),
        Nothing,
        CustomObj(Obj),
    }

    let items = vec![
        Item::Integer(1),
        Item::Text(String::from("apple")),
        Item::Array([3, 4]),
        Item::Vector(vec![1, 2, 3]),
        {
            let mut map = std::collections::HashMap::new();
            map.insert(String::from("key"), 1);
            Item::Map(map)
        },
        Item::Boolean(true),
        Item::Nothing,
        Item::CustomObj(Obj { x: 3 }),
    ];

    for item in items {
        print!("Value: {:?}, Type: ", item);
        match item {
            Item::Integer(n) => println!("Control expression is an integer: {}", n),
            Item::Text(s) => println!("Control expression is a string: {}", s),
            Item::Array(arr) => println!("Control expression is an array length 2: {:?}", arr),
            Item::Vector(vec) => println!("Control expression is a vector: {:?}", vec),
            Item::Map(map) => println!("Control expression is a map: {:?}", map),
            Item::Boolean(b) => println!("Control expression is a boolean: {}", b),
            Item::Nothing => println!("Control expression is Nothing"),
            Item::CustomObj(obj) => {
                if obj.x == 3 {
                    println!("Control expression is an Obj with x = 3")
                } else {
                    println!("Control expression is an Obj")
                }
            }
        }
    }


    
    println!("\n\n2 - How are the selectable segments specified?");
    let value = 5;
    match value {
        5 => println!("Case 5: Single statement"),
        10 => {
            println!("Case 10: Multiple statements");
            println!("You can have more than one line in a case block");
        }
        20 =>{
            println!("Case 20: single statement");
        }
        _ => (),
    }


    println!("\n\n3 - Is execution flow restricted to just a single selectable segment?");
    let value = 6;
    match value {
        2 | 4 | 6 => println!("Matched even number"),
        3 | 6 | 9 => println!("Matched divisible by 3"),
        _ => println!("No match"),
    }



    println!("\n\n4 - How are case values specified?");

    let value = 3;
    // literal values
    match value {
        1 => println!("Matched 1"),
        2 => println!("Matched 2"),
        3 => println!("Matched 3"),
        _ => (),
    }

    // variables
    let three = 2 + 1;
    let seven = 10 - 3;
    match value {
        three => println!("Matched 3"),
        seven => println!("Matched 7"),
        _ => (),
    }

    let strValue = "Hello";
    match strValue {
        "Hello" => println!("Matched Hello"),
        "World" => println!("Matched World"),
        _ => (),
    }

    // different types
    /*let value = 4;
    match value {
        "Hello" => println!("Different types are not allowed"),
        4 => println!("Different types are not allowed"),
        4.0 => println!("Different types are not allowed"),
        _ => (),
    }*/

    // multiple case values using OR patterns
    match value {
        1 | 2 | 3 => println!("Matched 1, 2, or 3"),
        4 | 5 | 6 => println!("Matched 4, 5, or 6"),
        _ => (),
    }

    // Pattern matching with ranges
    for i in 1..10 {
        match i {
            1..=3 => println!("Matched range 1 to 3: {}", i),
            4..=6 => println!("Matched range 4 to 6: {}", i),
            _ => println!("No match for: {}", i),
        }
    }
    
    // Type matching with enums
    #[derive(Debug)]
    enum Value {
        Int(i32),
        Str(String),
        Bool(bool),
    }

    let val = Value::Int(4);
    match val {
        Value::Int(i) => println!("Matched int: {}", i),
        Value::Str(s) => println!("Matched string: {}", s),
        Value::Bool(b) => println!("Matched bool: {}", b),
    }
    
    // expression with variables
    let m: i32 = 2;
    let n: i32 = 3;
    match value {
        x if x == m * n => println!("Matched 6"),
        x if x == m + n => println!("Matched 5"),
        x => println!("Matched other value: {}", x),
    }

    // variables with binding
    let a: i32 = 1;
    let b: i32 = 3;
    match value {
        x if x == a => println!("Matched a"),
        x if x == b => println!("Matched b"),
        _ => (),
    }


    println!("\n\n5 - What is done about unrepresented expression values?");
    let value = 100;
    match value {
        10 => println!("Matched 10"),
        _ => println!("No match for the given value (using wildcard)"),
    }
    match value {
        10 => println!("Matched 10"),
        default_value => println!("No match for value: {}", default_value),
    }
    /*match value {
        10 => println!("Matched 10"), // not compiled
    }*/
}
